maven-build-caching-samples/kotlin-project/README.md

```diff
@@ -0,0 +1,4 @@
+Example configuration for making the `compile` and `testCompile` goals of the `kotlin-maven-plugin` cacheable.
+
+Note that the incremental flag and output dir are ignored since it appears that the regular compiled output directory
```

Comment:
Doing `mvn clean` will wipe out everything, including the incremental output dir (at least the one configured by default). As far as I could experiment, the incremental compiler is relevant if you just run `mvn verify` without cleaning, it will compile only the changed source files and output them in both the incremental output dir and the regular output dir. See an example below:

```
└── target
    ├── classes
    │   ├── META-INF
    │   │   └── kotlin-project.kotlin_module
    │   └── com
    │       └── example
    │           └── AppKt.class
    ├── kotlin-ic
    │   ├── compile
    │   │   ├── build-history.bin
    │   │   ├── caches-jvm
    │   │   │   ├── inputs
    │   │   │   │   ├── source-snapshot.tab
    │   │   │   │   ├── [... lot more stuff...]
    │   │   │   ├── jvm
    │   │   │   │   └── kotlin
    │   │   │   │       ├── internal-name-to-source.tab
    │   │   │   │       ├── [... lot more stuff...]
    │   │   │   └── lookups
    │   │   │       ├── counters.tab
    │   │   ├── classes
    │   │   │   ├── META-INF
    │   │   │   │   └── kotlin-project.kotlin_module
    │   │   │   └── com
    │   │   │       └── example
    │   │   │           └── AppKt.class
    │   │   ├── last-build.bin
    │   │   └── snapshots.bin
```

In that case btw, the GE mvn extension does not attempt to use the cache since no clean has been issued.

As a summary:
- when `mvn clean verify`, takes advantage of getting cache hits for the regular compiled classes (thus ignore storing the incremental compilation dir/files into the cache)
- when `mvn verify`, takes advantage of the incremental compilation

Reply:
Thanks for checking that. 
@bigdaz what was tricky situation you saw with incremental compilation and the java mojo? can we test that here? Is it okay to ignore the incremental output folder as well?

Reply:
TL;DR I don't have any direct experience with true incremental compilation in Maven.

The Maven setting for _java_ incremental compile is `useIncrementalCompilation`. This is really misnamed:
- When `true` (the default) Maven will recompile _all_ source files in a module if any have changed
- When `false` (the broken option) Maven will only recompile changed source files, with no knowledge of source file dependencies.
See [here](https://stackoverflow.com/questions/16963012/maven-compiler-recompile-all-files-instead-modified) for more details.

In Gradle, we have tried to make caching work with incremental compilation by caching/restoring the analysis files along with the compiled classes. This works OK, except when compilation fails since we throw everything away. This was/is an issue for Snowflake since caching makes things slower as soon as a develop runs a build with a compilation failure.

Reply:
@alextu You might want to flag the incremental output dir as `localState`: this is a more accurate modelling of what it is.

Reply:
To be honest, I'm not sure what the Maven extension will do with properties labelled as `localState`. This would be good to have in the docs.

Reply:
As far as I understand it will just remove the corresponding `localState` folder(s) after unarchiving files from a cache hit. So in this particular case, it's useful only if the incremental dir is customized and outside of `target`, if it's the default it's just cleaned as part of the regular `clean` goal.